metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/external/impl/EntityClientResolver.java

```diff
@@ -0,0 +1,174 @@
+
+    private EntityClientApi entityClientApi;
+    private final int batchSize;
+
+
+    public EntityClientResolver(int batchSize) {
+        this.batchSize = batchSize;
+        entityClientApi = new EntityClientApiImpl();
+    }
+
+    @Override
+    public <T extends SearchTerm> Map<T, List<EnrichmentBase>> resolveByText(Set<T> searchTerms) {
+        final Function<T, EnrichmentQuery> inputFunction = EntityResolverUtils.createEnrichmentQueryForTextSearch();
+        return performInBatches(searchTerms, inputFunction);
+    }
+
+    @Override
+    public <T extends ReferenceTerm> Map<T, EnrichmentBase> resolveById(Set<T> referenceTerms) {
+        final Function<T, EnrichmentQuery> inputFunction = EntityResolverUtils.createEnrichmentQueryForRefSearch();
+        return EntityResolverUtils.pickFirstFromTheList(performInBatches(referenceTerms, inputFunction));
+    }
+
+    @Override
+    public <T extends ReferenceTerm> Map<T, List<EnrichmentBase>> resolveByUri(Set<T> referenceTerms) {
+        final Function<T, EnrichmentQuery> inputFunction = EntityResolverUtils.createEnrichmentQueryForRefSearch();
+        return performInBatches(referenceTerms, inputFunction);
+    }
+
+    /**
+     * Gets the Enrichment for Multiple input values in batches of batchSize
+     *
+     * @param inputValues Set of values for which enrichment will be performed
+     * @param requestParser Function to return EntityClientRequest from input value <I>
+     * @param <I> Input value Type. Could be <T extends SearchTerm> OR <T extends ReferenceTerm>
+     * @return
+     */
+    private <I> Map<I, List<EnrichmentBase>> performInBatches(Set<I> inputValues, Function<I, EnrichmentQuery> requestParser) {
+        final Map<I, List<EnrichmentBase>> result = new HashMap<>();
+        // create partitions
+        final List<List<I>> partitions = EntityResolverUtils.createBatch(inputValues, batchSize);
+
+        // Process partitions
+        for (List<I> partition : partitions) {
+            for (I input : partition) {
+                EnrichmentQuery clientRequest = requestParser.apply(input);
+                List<EnrichmentBase> enrichmentBaseList = executeEntityClientRequest(clientRequest);
+                if (!enrichmentBaseList.isEmpty()) {
+                    result.put(input, enrichmentBaseList);
+                }
+            }
+        }
+        return result;
+    }
+
+    private List<EnrichmentBase> executeEntityClientRequest(EnrichmentQuery clientRequest) {
+        // 1. get Entities
+        List<Entity> entities = getEntities(clientRequest);
+        if (!entities.isEmpty()) {
+            // 2. get the parent entities
+            List<Entity> parentEntities = new ArrayList<>();
+            entities.stream().forEach(entity -> {
+                if (entity.getIsPartOfArray() != null && EntityResolverUtils.isTextOrUriSearch(clientRequest)) {
+                    fetchParentEntities(entity, parentEntities);
```

Comment:
So this is done "remotely". In metis is done on the Persistent resolver which inside the service.
Could be quite a performance impact here.

Reply:
How about this one @SrishtiSingh-eu ?

Reply:
I had a thought of having a single class with all functionality. But if it will have a performance impact here. This could be moved to Service class

Reply:
When I say Service I don't mean the EnrichmentService class in metis code. I mean your Entity application service where the EntityClientApi point at.

Reply:
Okay. Hmmm now that I think of this, a method could be added to retrieve the Parents in entity API or EM. But as I discussed it with Hugo, we were under the impression to add the code in the metis. We should discuss this with Hugo if we want to move the code there.